### ISSUE TITLE
Please correct the bug in the code in io_monad/IO.scala...

### ISSUE THREAD

**Issue by chaotic3quilibrium** [alvinj/IOMonad#3]
Given I am literally using your book to learn about Monads, I was completely surprised to discover that io_monad/IO.scala was not correctly behaving in deferring execution. After investing a number of hours trying to figure out where I had messed up, I finally discovered it was a bug in your implementation.

It can be fixed in one of two ways. Either A) by fixing the current `def flatMap` method. Or B) by commenting that method out and uncommenting the `def flatMapOrig` method and renaming it to `flatMap`.

A) The `IO` must wrap the two result statements. By only wrapping the result of `result2` in `IO`, it is inadvertently and incorrectly eagerly executing `result1`. Here's a diff showing the change:
https://editor.mergely.com/v2T4PIdI/

B) I am curious why this was renamed and unused when it was the correct behavior. It would have saved me literally hours of time and confusion if there had been a comment indicating that it was a CORRECT implementation.
https://editor.mergely.com/EU6dX8xJ/

Tysvm for your book. With this one exception, I have deeply enjoyed it. And am looking forward to continuing my journey with it through the different Monads you explore later. And I really am enjoying you chose a very simple game context within which to help ground the concepts. It was exactly the simplicity I needed to grok what are fairly abstract and difficult concepts around which to wrap my head.

**Comment by tmichniewski** [comment]
Well, in my opinion the problem is even deeper and it is connect with the question of when the side effecting operation are really executed.

Please notice, that the `IO` class has by-name parameter, so it defers the computation of its parameter when the value gets originally lifted to IO monad - otherwise it could be executed even before the for expression (if we made for expression work on IO instances created before the for expression).

But then, the `flatMap` method cames and performs eagerly this series of operations whithout defering them. So, this implementation makes this whole `IO` monad behave eagerly (please notice that `map` method is also expressed in terms of `flatMap`). This is why in the PR for this repo someone claims that the explicit call to `run` is not necessary - yes, it is not necessary because this `IO` monad is eager and only pretends to be kind of lazy, but it is not.

However, if we left the original `flatMap` implementation (`flatMapOrig`), then the computation of `flatMap` and as a consequence `map` would behave as lazy computation. As a result, we would have to call the `run` method after the whole for expression to finally see the side effects.

We might see the same behaviour, if we replace the `map` implementation into something like:
```
def map[B](f: A => B): IO[B] = IO(f(a))
```
so got rid of explicit call to `flatMap`, which is eager. Then, in case of one generator, when no `flatMap` gets called, we would also see that we have to call the run method after the for expression to see the side effects, because then our monad would be kind of lazy.

Summing up - this `IO` monad is implemented a bit misleading, because we forgot about when the computation is going to be executed. And it can be executed:
- before the for expression if there are no call-by-name parameters and instances of `IO` monad are created before the for expression,
- in the middle of for expression, if the `flatMap` (and `map`) are eager, as right now,
- after the for expression, if we used `flatMapOrig`.

**Comment by alvinj** [comment]
Thank you (both) for this. I’m about 100 emails and issues behind at the moment, but I’ll take the time to look into this when I get into the `IO` monad as I’m updating *Functional Programming, Simplified”. In the meantime, many thanks!
